src/java/main/org/apache/zookeeper/server/NettyServerCnxn.java

```diff
@@ -87,6 +87,12 @@ public void close() {
             LOG.debug("close called for sessionid:0x"
                     + Long.toHexString(sessionId));
         }
+
+        // ZOOKEEPER-2743:
+        // Always unregister connection upon close to prevent
+        // connection bean leak under certain race conditions.
+        factory.unregisterConnection(this);
```

Comment:
Also it is guaranteed that the close call will be called at least once after the cnx bean is registered (e.g. when sever shutdown, the factor will iterate every cnx and close it..).